source/extensions/filters/common/expr/library/custom_library.cc

```diff
@@ -0,0 +1,109 @@
+                                   const Http::RequestHeaderMap* request_headers,
+                                   const Http::ResponseHeaderMap* response_headers,
+                                   const Http::ResponseTrailerMap* response_trailers) {
+  request_headers_ = request_headers;
+  response_headers_ = response_headers;
+  response_trailers_ = response_trailers;
+  // vocabulary
+  activation->InsertValueProducer(
+      CustomVocabularyName, std::make_unique<CustomVocabularyWrapper>(
+                                arena, info, request_headers, response_headers, response_trailers));
+  // Lazily evaluated functions only
+  absl::Status status;
+  status =
+      activation->InsertFunction(std::make_unique<GetDoubleCelFunction>(LazyEvalFuncNameGetDouble));
+  if (!status.ok()) {
+    ThrowException(LazyEvalFuncNameGetDouble, status);
+  }
+  status = activation->InsertFunction(
+      std::make_unique<GetProductCelFunction>(LazyEvalFuncNameGetProduct));
+  if (!status.ok()) {
+    ThrowException(LazyEvalFuncNameGetProduct, status);
+  }
+  status = activation->InsertFunction(std::make_unique<Get99CelFunction>(LazyEvalFuncNameGet99));
+  if (!status.ok()) {
+    ThrowException(LazyEvalFuncNameGet99, status);
+  }
+}
+
+void CustomLibrary::RegisterFunctions(CelFunctionRegistry* registry) const {
+  absl::Status status;
+  // lazily evaluated functions
+  status = registry->RegisterLazyFunction(
+      GetDoubleCelFunction::CreateDescriptor(LazyEvalFuncNameGetDouble));
+  if (!status.ok()) {
+    ThrowException(LazyEvalFuncNameGetDouble, status);
+  }
+  status = registry->RegisterLazyFunction(
+      GetProductCelFunction::CreateDescriptor(LazyEvalFuncNameGetProduct));
+  if (!status.ok()) {
+    ThrowException(LazyEvalFuncNameGetProduct, status);
+  }
+  status =
+      registry->RegisterLazyFunction(Get99CelFunction::CreateDescriptor(LazyEvalFuncNameGet99));
+  if (!status.ok()) {
+    ThrowException(LazyEvalFuncNameGet99, status);
+  }
+
+  // eagerly evaluated functions
+  status = google::api::expr::runtime::FunctionAdapter<CelValue, int64_t>::CreateAndRegister(
+      EagerEvalFuncNameGetNextInt, false, GetNextInt, registry);
+  if (!status.ok()) {
+    ThrowException(EagerEvalFuncNameGetNextInt, status);
+  }
+  status = google::api::expr::runtime::FunctionAdapter<CelValue, int64_t>::CreateAndRegister(
+      EagerEvalFuncNameGetSquareOf, true, GetSquareOf, registry);
+  if (!status.ok()) {
+    ThrowException(EagerEvalFuncNameGetSquareOf, status);
+  }
+}
+
+CustomLibraryPtr
+CustomLibraryFactory::createLibrary(const Protobuf::Message& config,
+                                    ProtobufMessage::ValidationVisitor& validation_visitor) {
+  const auto& typed_config =
+      MessageUtil::downcastAndValidate<const envoy::config::core::v3::TypedExtensionConfig&>(
+          config, validation_visitor);
+  const auto custom_library_config = MessageUtil::anyConvertAndValidate<CustomLibraryConfig>(
```

Comment:
I don't think individual extensions should need to do this kind of conversion. Instead, their factories should follow similar patterns as in other extensions. Maybe take a look at the templating that happens in various factory_base.h.